git-tool-providers/azure-dev-ops.mjs

```diff
@@ -1,33 +1,144 @@
+import { userMapping } from '../userMapping.mjs'
 
 const orgUrl = process.env.ORG_URL
 const token = process.env.AZURE_PERSONAL_ACCESS_TOKEN
-const repoId = process.env.REPO_ID
-const project = process.env.PROJECT_NAME
+const projects = process.env.PROJECT_NAMES.split(', ')
+const t = 'days between commits'
+const recent = `sum of all ${t}`
+const avgDays = `avg ${t}`
 
 export async function getAzureDevOpsProvider(orgUrl, token) {
-    let authHandler = azdev.getPersonalAccessTokenHandler(token); 
-    let webApi = new azdev.WebApi(orgUrl, authHandler);  
-    let gitApiObject = await webApi.getGitApi();
-    return gitApiObject;
+  let authHandler = azdev.getPersonalAccessTokenHandler(token)
+  let webApi = new azdev.WebApi(orgUrl, authHandler)
+  let gitApiObject = await webApi.getGitApi()
+
+  return gitApiObject
 }
 
 export async function getCommitersByName() {
-    let gitApiObject = await getAzureDevOpsProvider(orgUrl, token)
-    const commits = await gitApiObject.getCommits(repoId, {$skip: 0, $top: 15000}, project)
-    
-    if(commits && commits.length) {
-        let commiters = {}
-
-        commits.forEach((commit) => {
-            commiters[commit.author.name] ? 
-                commiters[commit.author.name].commitCount++ : commiters[commit.author.name] = { commitCount : 1 }
-        })
+  let gitApiObject = await getAzureDevOpsProvider(orgUrl, token)
+  const getCommits = async projectName => {
+    const repos = await gitApiObject.getRepositories(projectName)
+    const repoNames = repos.map(repo => repo.name)
+    return (
+      await Promise.all(
+        repoNames.map(repoName =>
+          gitApiObject.getCommits(
+            repoName,
+            {
+              fromDate: '01/01/2022 12:00:00 AM'
+            },
+            projectName,
+            0,
+            150000
+          )
+        )
+      )
+    ).flat()
+  }
+  const commits = (
+    await Promise.all(projects.map(project => getCommits(project)))
+  ).flat()
+  if (commits && commits.length) {
+    let commiters = {}
+    for (const commit of commits) {
+      if (commit) {
+        const author = commit?.author.name
+
+        const name = userMapping[author] ?? author
+        if (commiters[name]) {
+          const start = DateTime.fromJSDate(
+            commiters[name].mostRecentCommitDate
+          )
+          const end = DateTime.fromJSDate(commit.author.date)
 
-        return commiters
+          // find the interval between commit dates
+          commiters[name][recent] += start.diff(end, 'days').toObject().days
+          // running avg of the commit interval in days. done before count++ since intervals will always be one less than count.
+          commiters[name][avgDays] =
+            commiters[name][recent] / commiters[name].count
+
+          // iterate count
+          commiters[name].count++
+          // reassign date to current objects date
+          commiters[name].mostRecentCommitDate = commit.author.date
```

Comment:
this is only true if we continue to iterate through the commits in a date ordering? probably need to keep that in mind in case the getCommits ADO endpoint is every passed different search criteria

Reply:
The name is a misnomer, it is the most recent iteration commit date

Reply:
updated to be an array